extensions/amp-ad-network-adzerk-impl/0.1/amp-ad-network-adzerk-impl.js

```diff
@@ -0,0 +1,179 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  AmpA4A,
+  NO_CONTENT_RESPONSE,
+  CreativeMetaDataDef,
+} from '../../amp-a4a/0.1/amp-a4a';
+import {AmpAdTemplates} from '../../amp-a4a/0.1/amp-ad-templates';
+import {tryParseJson} from '../../../src/json';
+import {dev} from '../../../src/log';
+import {getMode} from '../../../src/mode';
+import {utf8Decode, utf8Encode} from '../../../src/utils/bytes';
+import {urls} from '../../../src/config';
+import {startsWith} from '../../../src/string';
+import {parseUrl} from '../../../src/url';
+
+/** @type {string} */
+const TAG = 'amp-ad-network-adzerk-impl';
+
+/** @visibleForTesting @type {string} */
+export const AMP_TEMPLATED_CREATIVE_HEADER_NAME = 'AMP-template-amp-creative';
+
+/** @typedef {{
+      ampCreativeTemplateUrl: string,
+      templateMacroValues: (JsonObject|undefined),
+    }} */
+let AmpTemplateCreativeDef;
+
+/** @private {?AmpAdTemplates} */
+let ampAdTemplates;
+
+/**
+ * Fast Fetch implementation for AdZerk network that allows AMP creative
+ * preferential render via AMP cache stored template expansion using
+ * amp-mustache.  AMP creative response will consist of the following JSON
+ * object with two fields:
+ *
+ * - ampCreativeTemplateId: number value for template ID.  Template must already
+ *    have been stored in the AMP cache.
+ * - templateMacroValues: optional JSON object mapping of macro name to its
+ *    string value used to dynamically update the template
+ *
+ * Additionally, ad response must include header indicating AMP creative
+ * template response: AMP-template-amp-creative: true
```

Comment:
instead of having the header value to be `true`, we can put `amp-mustache`, to differentiate templating languages.

Reply:
Done.